urlfcc351a1ba_kuntapavankumar_Z_tgz-pJ8-edu_concordia_dpis_ReplicaTestJ8/src/edu/concordia/dpis/Replica.java

```diff
@@ -0,0 +1,403 @@
+	 * election initiator process wins and becomes the new leader and sends the
+	 * COORDINATOR message to every other process. If one of the higher
+	 * processes number (p5 or p6) replies, then they takes over and p4's work
+	 * is done. When p5 or p6 receives an ELECTION message from one of its lower
+	 * numbered processes they send an OK message to the sender p4. The receiver
+	 * (p5 and p6) holds an election if it is not conducting one. As it goes on
+	 * p5 gives up and the one remaining process (p6) is the new group leader.
+	 * P6 announces to every other process that the new leader is in effect
+	 * immediately.
+	 */
+	@Override
+	public MessageType election(String replicaId) {
+		System.out.println("election is going to be started by " + replicaId);
+		// if the election is started by a lower id.
+		if (Integer.parseInt(this.address.getId()) >= Integer
+				.parseInt(replicaId) && !isElectionInProgress) {
+			leaderName = null;
+			new Thread(new Runnable() {
+				@Override
+				public void run() {
+					for (final Node node : nodes) {
+						if (leaderName == null) {
+							try {
+								// get all nodes whose id's are greater than
+								// this replica's id
+								if (Integer.parseInt(node.getAddress().getId()) > Integer
+										.parseInt(address.getId())) {
+									System.out
+											.println("sent election message to node deployed on "
+													+ node.getAddress()
+															.getHost()
+													+ " and on port "
+													+ node.getAddress()
+															.getPort());
+									// send an election message
+									MessageType mType = node.election(address
+											.getId());
+									// incase the node replied with a
+									// COORDINATOR message, that it being the
+									// leader
+									if (MessageType.COORDINATOR.equals(mType)) {
+										System.out
+												.println("node deployed on "
+														+ node.getAddress()
+																.getHost()
+														+ " and on port "
+														+ node.getAddress()
+																.getPort()
+														+ " replied with a COORDINATOR message");
+										// update the new leader info in this
+										// replica
+										newLeader(node.getAddress().getId());
+										// or if the node replied with an OK
+										// message, meaning it takes care of the
+										// next step in election
+									} else if (MessageType.OK.equals(mType)) {
+										System.out.println("node deployed on "
+												+ node.getAddress().getHost()
+												+ " and on port"
+												+ node.getAddress().getPort()
+												+ " replied with a OK message");
+										try {
+											// and this replica calms down and
+											// waits for the leader emerge
+											System.out
+													.println("will wait for some time to let some one inform me about the new leader");
+											Thread.sleep((long) (3000 + (1000 * Math
+													.random())));
+										} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)